### PR TITLE
Refactor2

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
This PR should be reviewed after #28

- Some boost stuff replaced by std library
- Added an overview of classes in core nodes
- Moved most of implementation of `rgbd_to_ros` to ServerROS class
- Add `ros::master::check()` to all nodes

Merge together with tue-robotics/robot_launch_files#57

### CODE REVIEW

**Comment by MatthijsBurgh** [comment]
@jlunenburg Ready for review

File: rgbd/README.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ It provides a unified format for depthmaps combined with RGB data, which is more
 All **bold names** are classes. The main interface types are shown to understand the differences between all classes.
 ![core classes](doc/rgbd_classes.svg)
 
+## Core nodes
```
Comment (jlunenburg) [inline rgbd/README.md:11]: Move figure to description

File: rgbd/src/get_3d_point_from_image_roi_node.cpp [hunk 2]
```diff
@@ -130,7 +132,7 @@ int main(int argc, char **argv)
     ros::Time last_image_stamp;
 
     ros::Rate r(max_fps);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/src/get_3d_point_from_image_roi_node.cpp:135]: If `ros::master::check()` does not return true, something's wrong and I'd like to see an error message (and `exit(1)`)

File: rgbd/src/rgbd_to_ros.cpp [hunk 3]
```diff
@@ -1,134 +1,98 @@
...
+            {
+                publish_rgb = true;
+                publish_depth = true;
+                valid_arg_provided = true;
+            }
+            else if (opt == "--pc" || opt == "--pointcloud")
+            {
+                publish_pc = true;
+                valid_arg_provided = true;
+            }
```
Comment (jlunenburg) [inline rgbd/src/rgbd_to_ros.cpp:60]: Add an 'else' in case the provided argument isn't known

File: rgbd/src/rgbd_to_ros.cpp [hunk 4]
```diff
@@ -1,134 +1,98 @@
...
+        }
+        if (!valid_arg_provided)
+        {
+            publish_rgb = true;
+            publish_depth = true;
+        }
     }
 
+    ROS_DEBUG_STREAM("publish_rgb: " << publish_rgb);
```
Comment (jlunenburg) [inline rgbd/src/rgbd_to_ros.cpp:69]: After init and start

File: rgbd/src/rgbd_to_ros.cpp [hunk 5]
```diff
@@ -1,134 +1,98 @@
...
-    {
-        if (!last_image_stamp.isZero() && ros::Time::now() - last_image_stamp > ros::Duration(5.0))
-        {
-          ROS_ERROR("rgbd to ros did not receive images for 5 seconds ... restarting ...");
-          exit(1);
-        }
+    rgbd::Image image;
 
-        rgbd::Image image;
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/src/rgbd_to_ros.cpp:91]: `ros::master::check` as in previous file

File: rgbd/src/rgbd_to_shm.cpp [hunk 6]
```diff
@@ -31,7 +32,7 @@ int main(int argc, char **argv) {
     rgbd::ImagePtr image_ptr;
 
     ros::Rate r(max_fps);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/src/rgbd_to_shm.cpp:35]: Idem

File: rgbd/src/ros_to_rgbd.cpp [hunk 7]
```diff
@@ -65,7 +67,7 @@ int main(int argc, char **argv) {
     rgbd::ImagePtr image_ptr;
 
     ros::Rate r(max_fps);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/src/ros_to_rgbd.cpp:70]: Idem

File: rgbd/src/server_ros.cpp [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+void ServerROS::send(const Image& image)
+{
+    if ((pub_depth_img_ || pub_depth_pc_) && image.getDepthImage().data)
+    {
+        // Convert camera info to message
+        rgbd::View view(image, image.getDepthImage().cols);
+
+
+        if (pub_depth_img_ && (pub_depth_img_->getNumSubscribers() || pub_depth_img_->getNumSubscribers()))
```
Comment (jlunenburg) [inline rgbd/src/server_ros.cpp:82]: * `pub_depth_img_ was already checked`
* `|| pub_depth_info...`

File: rgbd/src/server_ros.cpp [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+// ----------------------------------------------------------------------------------------
+
+void ServerROS::send(const Image& image)
+{
+    if ((pub_depth_img_ || pub_depth_pc_) && image.getDepthImage().data)
+    {
+        // Convert camera info to message
+        rgbd::View view(image, image.getDepthImage().cols);
+
```
Comment (jlunenburg) [inline rgbd/src/server_ros.cpp:80]: Whitelines

File: rgbd/test/test_client.cpp [hunk 10]
```diff
@@ -21,7 +22,7 @@ int main(int argc, char **argv) {
     rgbd::Image image;
 
     ros::Rate r(rate);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/test/test_client.cpp:25]: Idem

File: rgbd/test/test_server.cpp [hunk 11]
```diff
@@ -39,7 +40,7 @@ int main(int argc, char **argv)
     rgbd::Image image(rgb_image, depth_image, cam_model, "test_frame_id", ros::Time::now().toSec());
 
     int x = 0;
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/test/test_server.cpp:43]: Idem

File: rgbd/tools/analyser.cpp [hunk 12]
```diff
@@ -13,10 +19,10 @@ int main(int argc, char **argv)
     ros::init(argc, argv, "rgbd_viewer");
 
     rgbd::Client client;
-    client.intialize(argv[1]);
+    client.intialize(ros::names::resolve(argv[1]));
 
     ros::Rate r(30);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/tools/analyser.cpp:25]: Idem

File: rgbd/tools/full_screen_viewer.cpp [hunk 13]
```diff
@@ -22,7 +23,7 @@ int main(int argc, char **argv)
     cv::Mat canvas;
 
     ros::Rate r(30);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/tools/full_screen_viewer.cpp:26]: Idem

File: rgbd/tools/multitool.cpp [hunk 14]
```diff
@@ -103,7 +104,7 @@ int main(int argc, char **argv)
     rgbd::ImagePtr image;
 
     ros::Rate r(30);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/tools/multitool.cpp:107]: Idem

File: rgbd/tools/presenter.cpp [hunk 15]
```diff
@@ -36,7 +42,7 @@ int main(int argc, char **argv)
 
     cv::Mat canvas(600, 1024, CV_8UC3, cv::Scalar(0, 0, 0));
 
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/tools/presenter.cpp:45]: Idem

File: rgbd/tools/presenter.cpp [hunk 16]
```diff
@@ -1,6 +1,12 @@
+#include <opencv2/highgui/highgui.hpp>
+
```
Comment (jlunenburg) [inline rgbd/tools/presenter.cpp:2]: Do we still need this?

File: rgbd/tools/record_to_video.cpp [hunk 17]
```diff
@@ -73,7 +75,7 @@ int main(int argc, char **argv) {
 
     // Start loop at given frequency
     ros::Rate r(video_frame_rate);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/tools/record_to_video.cpp:78]: Idem

File: rgbd/tools/rgbd_to_rgb_png.cpp [hunk 18]
```diff
@@ -14,7 +14,7 @@ int main(int argc, char **argv) {
         return 1;
     }
 
-    for (unsigned int i = 1; i < argc; ++i)
+    for (int i = 1; i < argc; ++i)
```
Comment (jlunenburg) [inline rgbd/tools/rgbd_to_rgb_png.cpp:17]: Is `argc` signed?

File: rgbd/tools/save.cpp [hunk 19]
```diff
@@ -19,7 +20,7 @@ int main(int argc, char **argv) {
     rgbd::Image image;
 
     ros::Rate r(30);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/tools/save.cpp:23]: Idem

File: rgbd/tools/viewer.cpp [hunk 20]
```diff
@@ -14,7 +17,7 @@ int main(int argc, char **argv)
     client.intialize(ros::names::resolve("rgbd"));
 
     ros::Rate r(30);
-    while (ros::ok())
+    while (ros::ok() && ros::master::check())
```
Comment (jlunenburg) [inline rgbd/tools/viewer.cpp:20]: Idem

File: rgbd/tools/rgbd_to_rgb_png.cpp (hunk 18 above)
Comment (MatthijsBurgh) [reply rgbd/tools/rgbd_to_rgb_png.cpp:17]: yes it is.

**Comment by MatthijsBurgh** [comment]
I have tested all of the core nodes. And all ways of communication between server and client.

**Pull request merged**